pkg/block/azure/adapter.go

```diff
@@ -270,35 +271,57 @@ func (a *Adapter) getPreSignedURL(ctx context.Context, obj block.ObjectPointer,
 		return "", err
 	}
 
-	// Use shared credential for clients initialized with storage access key
-	if qualifiedKey.StorageAccountName == a.clientCache.params.StorageAccount && a.clientCache.params.StorageAccessKey != "" {
-		container, err := a.clientCache.NewContainerClient(qualifiedKey.StorageAccountName, qualifiedKey.ContainerName)
-		if err != nil {
-			return "", err
-		}
-		client := container.NewBlobClient(qualifiedKey.BlobURL)
-		urlExpiry := a.newPreSignedTime()
-		return client.GetSASURL(permissions, urlExpiry, &blob.GetSASURLOptions{})
+	// Add content-disposition if filename provided
+	var contentDisposition string
+	if permissions.Read && filename != "" {
+		contentDisposition = mime.FormatMediaType("attachment", map[string]string{"filename": path.Base(filename)})
 	}
 
-	// Otherwise assume using role based credentials and build signed URL using user delegation credentials
-	urlExpiry := a.newPreSignedTime()
-	udc, err := a.clientCache.NewUDC(ctx, qualifiedKey.StorageAccountName, &urlExpiry)
+	// Snapshot time
+	urlParts, err := sas.ParseURL(qualifiedKey.BlobURL)
 	if err != nil {
 		return "", err
 	}
+	snapshotTime, err := time.Parse(blob.SnapshotTimeFormat, urlParts.Snapshot)
+	if err != nil {
+		snapshotTime = time.Time{}
+	}
 
 	// Create Blob Signature Values with desired permissions and sign with user delegation credential
+	urlExpiry := a.newPreSignedTime()
 	blobSignatureValues := sas.BlobSignatureValues{
-		Protocol:      sas.ProtocolHTTPS,
-		ExpiryTime:    urlExpiry,
-		Permissions:   to.Ptr(permissions).String(),
-		ContainerName: qualifiedKey.ContainerName,
-		BlobName:      qualifiedKey.BlobURL,
+		Protocol:           sas.ProtocolHTTPS,
+		ExpiryTime:         urlExpiry,
+		Version:            sas.Version,
+		Permissions:        permissions.String(),
+		ContainerName:      qualifiedKey.ContainerName,
+		BlobName:           qualifiedKey.BlobURL,
+		ContentDisposition: contentDisposition,
+		SnapshotTime:       snapshotTime,
 	}
-	sasQueryParams, err := blobSignatureValues.SignWithUserDelegation(udc)
-	if err != nil {
-		return "", err
```

Comment:
Cool, but just to be on the safe side, how was this tested?

Reply:
for s3, azure, and gs: used lakectl fs upload/download/stats using --pre-sign.
using the UI upload and copy presigned url + paste into the browser.